contracts/src/registry/PermissionedRegistry.sol

```diff
@@ -2,55 +2,269 @@
+    uint256 private constant ROLE_SET_SUBREGISTRY = 1 << 2;
+    uint256 private constant ROLE_SET_SUBREGISTRY_ADMIN = ROLE_SET_SUBREGISTRY << 128;
+
+    uint256 private constant ROLE_SET_RESOLVER = 1 << 3;
+    uint256 private constant ROLE_SET_RESOLVER_ADMIN = ROLE_SET_RESOLVER << 128;
+
+    uint256 private constant ROLE_SET_TOKEN_OBSERVER = 1 << 4;
+    uint256 private constant ROLE_SET_TOKEN_OBSERVER_ADMIN = ROLE_SET_TOKEN_OBSERVER << 128;
+
+    modifier onlyNonExpiredTokenRoles(uint256 tokenId, uint256 roleBitmap) {
+        _checkRoles(tokenIdResource(tokenId), roleBitmap, _msgSender());
+        (, uint64 expires, ) = datastore.getSubregistry(tokenId);
+        if (expires < block.timestamp) {
+            revert NameExpired(tokenId);
+        }
+        _;
     }
 
-    function _setFlags(uint256 tokenId, uint96 _flags)
-        internal
-        withSubregistryFlags(tokenId, FLAG_FLAGS_LOCKED, 0)
-        returns(uint96 newFlags)
+    constructor(IRegistryDatastore _datastore, IRegistryMetadata _metadata) BaseRegistry(_datastore) MetadataMixin(_metadata) {
+        _grantRoles(ROOT_RESOURCE, ALL_ROLES, _msgSender(), false);
+
+        if (address(_metadata) == address(0)) {
+            _updateMetadataProvider(new SimpleRegistryMetadata());
+        }
+    }
+
+    function uri(uint256 tokenId ) public view override returns (string memory) {
+        return tokenURI(tokenId);
+    }
+
+    function ownerOf(uint256 tokenId)
+        public
+        view
+        virtual
+        override(ERC1155Singleton, IERC1155Singleton)
+        returns (address)
     {
-        (address subregistry, uint96 oldFlags) = datastore.getSubregistry(tokenId);
-        newFlags = oldFlags | (_flags & FLAGS_MASK);
-        if (newFlags != oldFlags) {
-            datastore.setSubregistry(tokenId, subregistry, newFlags);
+        (, uint64 expires, ) = datastore.getSubregistry(tokenId);
+        if (expires < block.timestamp) {
+            return address(0);
         }
+        return super.ownerOf(tokenId);
+    }
+
+    function register(string calldata label, address owner, IRegistry registry, address resolver, uint256 roleBitmap, uint64 expires)
+        public
+        onlyRootRoles(ROLE_REGISTRAR)
+        returns (uint256 tokenId)
+    {
+        tokenId = NameUtils.labelToTokenId(label);
+
+        (, uint64 oldExpiry, uint32 tokenIdVersion) = datastore.getSubregistry(tokenId);
+
+        if (oldExpiry >= block.timestamp) {
+            revert NameAlreadyRegistered(label);
+        }
+
+        if (expires < block.timestamp) {
+            revert CannotSetPastExpiration(expires);
+        }
+
+        tokenId = _constructVersionedTokenId(tokenId, tokenIdVersion);
+
+        // if there is a previous owner, burn the token
+        address previousOwner = super.ownerOf(tokenId);
+        if (previousOwner != address(0)) {
+            _burn(previousOwner, tokenId, 1);
+            tokenId = _regenerateTokenId(tokenId); // so we have a fresh acl
```

Comment:
We already know the `registry` and `expires` here; can we pass them in to this function to avoid re-fetching them?